shakenfist/external_api/instance.py

```diff
@@ -135,19 +136,35 @@ def post(self, name=None, cpus=None, memory=None, network=None, disk=None,
                     Artifact.TYPE_LABEL, '%s%s/%s' % (LABEL_URL, get_jwt_identity(), label))
                 if not a:
                     return api_base.error(404, 'label %s not found' % label)
+                if a.state.value != Artifact.STATE_CREATED:
+                    return api_base.error(404, 'label %s is not ready (state=%s)'
+                                          % (label, a.state.value))
                 blob_uuid = a.most_recent_index.get('blob_uuid')
                 if not blob_uuid:
                     return api_base.error(404, 'label %s not found (no versions)' % label)
+                b = Blob.from_db(blob_uuid)
+                if not b:
+                    return api_base.error(404, 'artifact references non-existent blob (%s)' % blob_uuid)
+                if b.state == Blob.STATE_DELETED:
+                    return api_base.error(404, 'artifact references deleted blob (%s)' % blob_uuid)
```

Comment:
In general elsewhere we don't tell them _why_ the object isn't found. Just that its not found. I guess I was thinking that we shouldn't confirm that the object ever existed? Thoughts? Maybe my concern applies more to the "incorrect permissions" case.

Reply:
I agree in the security case. In this case, I'm thinking that the user might have a blob UUID and attempt to use it after deletion. This msg explains why a previously valid blob UUID does not work.

Being UUID's, it's unlikely they will guess a foreign UUID. It also does not reveal any information about the blob.

When Blobs are namespace'd, the security check should be before this message.